Cleanup of `eat_while()` in lexer

Julian-Wollersberger opened the pull request:
The size of a lexer Token was inflated by the largest `TokenKind` variants `LiteralKind::RawStr` and `RawByteStr`, because 
* it used `usize` although `u32` is sufficient in rustc, since crates must be smaller than 4GB,
* and it stored the 20 bytes big `RawStrError` enum for error reporting.

If a raw string is invalid, it now needs to be reparsed to get the `RawStrError` data, but that is a very cold code path.

Technically this breaks other tools that depend on rustc_lexer because they are now also restricted to a max file size of 4GB. But this shouldn't matter in practice, and rustc_lexer isn't stable anyway.

Can I also get a perf run?

Edit: This makes no difference in performance. The PR now only contains a small cleanup.

rust-highfive commented:
r? @varkor

(rust_highfive has picked a reviewer for you, use r? to override)

jonas-schievink commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 2cc461ba86901295d9b706d7e39a15e3d4efa6cf with merge 42c8db8af9dcb6f65f03cbbdf104da7ed2fc2018...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2cc461ba86901295d9b706d7e39a15e3d4efa6cf","merge_sha":"42c8db8af9dcb6f65f03cbbdf104da7ed2fc2018"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

pickfire commented on compiler/rustc_lexer/src/cursor.rs:
```diff
@@ -61,8 +62,8 @@ impl<'a> Cursor<'a> {
-    pub(crate) fn len_consumed(&self) -> usize {
-        self.initial_len - self.chars.as_str().len()
+    pub(crate) fn len_consumed(&self) -> u32 {
+        (self.initial_len - self.chars.as_str().len()).try_into().unwrap()
```
Would this be clearer?
```suggestion
        let len = self.initial_len - self.chars.as_str().len();
        len.try_into().unwrap()
```
I wonder do we just unwrap if the integer overflow?

Julian-Wollersberger replied:
Changed. 
Yes, it panics on overflow. It was suggested by the compiler and I guess it's more explicit than `as u32`.

Julian-Wollersberger commented:
I fixed the test failure and added @pickfire's suggestion.

I don't understand why bors timed out. Was it spurious?
Can I get another perf run?

Xanewok commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 2663c35bf86871e1849400eca6417c1fed02990c with merge b468e924ec58d34e078526586335258b57deb9a8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2663c35bf86871e1849400eca6417c1fed02990c","merge_sha":"b468e924ec58d34e078526586335258b57deb9a8"} -->

Julian-Wollersberger commented:
I noticed a potential bug in `eat_while()`: it doesn't account for the number of UTF8 bytes.
Fixed it by inlining it in the two places where the count is used and also simplified the logic there. 
It might be a little perf improvement too.

@Xanewok can you restart the perf-run? Sorry for bothering...

jyn514 commented:
@Julian-Wollersberger bug fixes don't need a perf run IMO, unless they strongly affect perf.

varkor commented on compiler/rustc_parse/src/lexer/mod.rs:
```diff
@@ -447,17 +451,17 @@ impl<'a> StringReader<'a> {
-            Some(RawStrError::InvalidStarter { bad_char }) => {
+    fn report_raw_str_error(&self, start: BytePos, prefix_len: u32) {
+        let src = &self.src[start.to_usize() + prefix_len as usize..];
+        let err = rustc_lexer::recompute_raw_string_error(src, prefix_len);
```
Is there a reason we can't use inline `recompute_raw_string_error` if it's only used in this place?

Julian-Wollersberger replied:
Because `Cursor` and its impls are private in rustc_lexer, so I made that wrapper.

pickfire commented on compiler/rustc_lexer/src/lib.rs:
```diff
@@ -38,11 +38,11 @@ use std::convert::TryFrom;
 pub struct Token {
     pub kind: TokenKind,
-    pub len: usize,
+    pub len: u32,
```
I bet we could use a smaller value for token, like `u16` since the minimum file size is `u32` fat32 4GiB, I wonder if this reduce the size of `Token`?

Julian-Wollersberger replied:
No, because someone will need a string literal with more than 65 535 chars for example. Then u16 would overflow.

pickfire replied:
I know, but I think even the editor cannot open it with 65535 characters in it or can it? Is it even possible to see there be a 65535 long string literal?

Julian-Wollersberger replied:
Yes. It's not good style, but possible. Think about proc macros that put the content of a file into a string. 
Or a block comment can easily span a file with thousands of lines.

Somebody one day will need to hack together a way to put large binary files into the source code, and they will even hit the 4GB limit. Ideally we'd make the lengths `u64` but that is a performance tradeoff.

pickfire replied:
Oh, I forgot comments is part of the token. Indeed comments is the one that could easily break the limit unlike the rest of the usual tokens, I was only thinking of the strings, numbers and such but not comments.

pickfire commented on compiler/rustc_lexer/src/cursor.rs:
```diff
@@ -61,8 +62,9 @@ impl<'a> Cursor<'a> {
-    pub(crate) fn len_consumed(&self) -> usize {
-        self.initial_len - self.chars.as_str().len()
+    pub(crate) fn len_consumed(&self) -> u32 {
+        let len = self.initial_len - self.chars.as_str().len();
```
Should we make `initial_len` `u32`?

Julian-Wollersberger replied:
That wouldn't reduce the number of `as usize` or `as u32` needed.

pickfire replied:
Yes, but I don't see why we need `initial_len` to be `usize` when the total len is `u16`, unless the `usize` for the computer is `u8` but otherwise it wouldn't happen right?

Julian-Wollersberger replied:
We don't need it, but it would be more annoying to implement and not be a perf improvement anyway.

pickfire commented on compiler/rustc_lexer/src/lib.rs:
```diff
@@ -151,24 +151,27 @@ pub enum LiteralKind {
+    RawStr { n_hashes: u16, valid: bool },
     /// "br"abc"", "br#"abc"#", "br####"ab"###"c"####", "br#"a"
-    RawByteStr { n_hashes: u16, err: Option<RawStrError> },
+    RawByteStr { n_hashes: u16, valid: bool },
```
Could `n_hashes` be `NonZeroU16` to reduce size such that `bool` could be a niche?

jyn514 replied:
No, because `r"a"` is a valid number of hashes.

pickfire replied:
Why not make it `u8`? Would people go crazy and put 255 `#`?

Julian-Wollersberger replied:
It was decided agains making num hashes `u8` in #50296. And tricks like adding one makes the code harder to maintain.
I looked into making `Token` even smaller, but that would need many more adjustments. I guess 8 bytes instead of 16 would be possible, but very complicated.

pickfire replied:
Nice findings, but @nikomatsakis said it was ok https://github.com/rust-lang/rust/pull/50296#issuecomment-385464881.

The reason why `u16` was used is because it was thought that `u8` brings no benefits over `u16`. But here it does bring a benefit, that is reducing the size of token to **8** bytes (without this would be 10 with the other changes I mentioned) which could probably improve compilation time a bit. Maybe CC @rust-lang/lang

scottmcm replied:
> Could n_hashes be NonZeroU16 to reduce size such that bool could be a niche?

`(NonZeroU16, bool)` can't space-optimize anyway.

pickfire replied:
Yeah, `NonZero` variant can't space optimize here but `u8` could.

pickfire commented on compiler/rustc_lexer/src/lib.rs:
```diff
@@ -151,24 +151,27 @@ pub enum LiteralKind {
     /// The string was never terminated. `possible_terminator_offset` is the number of characters after `r` or `br` where they
     /// may have intended to terminate it.
-    NoTerminator { expected: usize, found: usize, possible_terminator_offset: Option<usize> },
+    NoTerminator { expected: u32, found: u32, possible_terminator_offset: Option<u32> },
```
Could this be `u16`? I think it is hard for raw string to hit `u32` in source code which editor would have a hard time handling it? Or would it? I think this is a less safe choice compared to the reductions I mentioned above but looks possible.

Julian-Wollersberger replied:
It doesn't matter here because `RawStrError`s aren't stored anymore and aren't constructed that often. I also didn't want to regress potential errors in weird cases.

pickfire commented on compiler/rustc_lexer/src/lib.rs:
```diff
@@ -682,14 +685,19 @@ impl Cursor<'_> {
         let start_pos = self.len_consumed();
         let mut possible_terminator_offset = None;
         let mut max_hashes = 0;
 
         // Count opening '#' symbols.
-        let n_start_hashes = self.eat_while(|c| c == '#');
+        let mut eaten = 0;
+        while self.first() == '#' {
+            eaten += 1;
+            self.bump();
+        }
+        let n_start_hashes = eaten;
```
Is this necessary?

Julian-Wollersberger commented:
I guess rust-timer got stuck. @Mark-Simulacrum do you know what happened here and can you restart the perf run?
Maybe rust-timer got confused because bors timed out?

jyn514 commented:
@rust-timer queue

rust-timer commented:
Awaiting bors try build completion

jyn514 commented:
🤦 I think it got confused because you pushed more commits maybe?

@bors try

bors commented:
:hourglass: Trying commit 8caa6072eef69d6f4098ed7a04f2e674c22dbe57 with merge 85f2a07ff3f2a86b9ef4ebe097dc23656933e80b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"8caa6072eef69d6f4098ed7a04f2e674c22dbe57","merge_sha":"85f2a07ff3f2a86b9ef4ebe097dc23656933e80b"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1226323031), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37105)
Build commit: 85f2a07ff3f2a86b9ef4ebe097dc23656933e80b (`85f2a07ff3f2a86b9ef4ebe097dc23656933e80b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37105","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1226323031"},"merge_sha":"85f2a07ff3f2a86b9ef4ebe097dc23656933e80b"} -->

rust-timer commented:
Queued 85f2a07ff3f2a86b9ef4ebe097dc23656933e80b with parent f1dab244d798cda027a327fa43a17547d489511d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=f1dab244d798cda027a327fa43a17547d489511d&end=85f2a07ff3f2a86b9ef4ebe097dc23656933e80b).

rust-timer commented:
Finished benchmarking try commit (85f2a07ff3f2a86b9ef4ebe097dc23656933e80b): [comparison url](https://perf.rust-lang.org/compare.html?start=f1dab244d798cda027a327fa43a17547d489511d&end=85f2a07ff3f2a86b9ef4ebe097dc23656933e80b).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

Julian-Wollersberger commented:
The perf results couldn't be more neutral. Oh well.
I've removed the first two commits and only the bugfix remains. This can be `rollup=always` I guess.

Changing `usize` to `u32` might still be worth it as a cleanup so rustc_lexer is more consistent with the rest of rustc. Opinions? 
@varkor this is ready for review again.

pickfire commented:
Oh, there doesn't seemed to be performance gain but I wonder if memory footprint is reduced? From what I see the memory footprint is not tracked in the benchmarks.

Mark-Simulacrum commented:
Please update the PR description and title if you significantly change the pull request, right now it is not clear that this is a bugfix. I would also expect to see a test case if there is a bug being fixed here.

mati865 commented:
> 
> 
> Oh, there doesn't seemed to be much performance gain but I wonder if memory footprint is reduced? From what I see the memory footprint is not tracked in the benchmarks.

It's noisy as always but in terms of the memory it's neutral/slightly worse.

Julian-Wollersberger commented:
> Please update the PR description and title if you significantly change the pull request
Ok, makes sense. Done.

> I would also expect to see a test case if there is a bug being fixed here.
Sorry, I wasn't precise. It _could_ have been a bug if the count returned by `eat_while()` was used in a place where non-ASCII chars can be, but luckily it was only used to count the '#' in raw strings, which are ASCII chars.
So it's not really a bugfix but a cleanup to prevent a potential bug.

Julian-Wollersberger commented:
> From what I see the memory footprint is not tracked in the benchmarks.

@pickfire At the bottom of the page you can select "max-rss" and click "submit" at the right edge of the page.
(There is PR rust-lang/rustc-perf#782 to improve the design.)

varkor commented:
@bors r+ rollup

bors commented:
:pushpin: Commit bd49ded308f7243d1ba3170ea1bd0d5855d0544b has been approved by `varkor`

<!-- @bors r=varkor bd49ded308f7243d1ba3170ea1bd0d5855d0544b -->
<!-- homu: {"type":"Approved","sha":"bd49ded308f7243d1ba3170ea1bd0d5855d0544b","approver":"varkor"} -->
